## Baseline Model Results

The supplied CNN model was executed on the chest X-ray dataset containing three classes: BACTERIAL, NORMAL, and VIRAL. The dataset loader reported 4336 training images, 1083 validation images, and 437 test images.

The baseline model was trained for 8 epochs. The final training accuracy reached 0.8808, while the final validation accuracy was 0.7636. Test accuracy was 0.7597.

The accuracy curves suggest signs of overfitting, because training accuracy continued to improve while validation accuracy stopped improving and declined slightly in later epochs.

## Training Time

The CNN model was trained for 8 epochs on the chest X-ray dataset.

The total training time was approximately 150.14 seconds.

This equates to roughly 18–20 seconds per epoch.

The model achieved a test accuracy of approximately 0.75 (75.05%).

In [1]:
import os
import pandas as pd

train_path = r'F:\ComputerVision_Assignment2\chest_xray\train'
test_path = r'F:\ComputerVision_Assignment2\chest_xray\test'

def count_images(folder):
    counts = {}
    for class_name in os.listdir(folder):
        class_folder = os.path.join(folder, class_name)
        if os.path.isdir(class_folder):
            counts[class_name] = len([
                f for f in os.listdir(class_folder)
                if os.path.isfile(os.path.join(class_folder, f))
            ])
    return counts

train_counts = count_images(train_path)
test_counts = count_images(test_path)

print("Train:", train_counts)
print("Test:", test_counts)

Train: {'BACTERIAL': 2596, 'NORMAL': 1461, 'VIRAL': 1363}
Test: {'BACTERIAL': 184, 'NORMAL': 122, 'VIRAL': 132}


## Dataset Distribution

The dataset consists of three classes: BACTERIAL, NORMAL, and VIRAL.

The training set contains:
- BACTERIAL: 2596 images
- NORMAL: 1461 images
- VIRAL: 1363 images

The test set contains:
- BACTERIAL: 184 images
- NORMAL: 122 images
- VIRAL: 132 images

The dataset is imbalanced, as the BACTERIAL class contains significantly more samples than the NORMAL and VIRAL classes.

This imbalance may cause the model to bias towards the majority class, reducing performance on minority classes.

To address this issue, techniques such as class weighting or data augmentation can be used.

## Overfitting Analysis

The model shows signs of overfitting.

Training accuracy increases steadily from approximately 0.72 to 0.88 over the 8 epochs. However, validation accuracy peaks at around 0.78 and then begins to decrease slightly in later epochs.

This indicates that the model is learning the training data well but is not generalising effectively to unseen validation data.

The gap between training and validation accuracy suggests that the model is overfitting.

To address overfitting, techniques such as using GlobalAveragePooling2D instead of Flatten, adding regularisation, or applying data augmentation can be used.

## Data Augmentation

To reduce overfitting, data augmentation techniques were applied.

Specifically, random horizontal flipping and small rotations were introduced to the training images.

This increases the diversity of the dataset and helps the model generalise better.

After applying data augmentation, the model was retrained.

The test accuracy decreased slightly from 0.7597 to 0.7345.

However, the validation accuracy became more stable, indicating reduced overfitting.

This suggests that the model is learning more general features rather than memorising the training data.